nightingale_manipulation/src/nightingale_manipulation/manipulation_action_client.py

```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/env python3
+
+import actionlib
+import rospy
+
+# joint
+from moveit_action_handlers.msg import PropertyValuePair
+from moveit_action_handlers.msg import MoveToJointsMoveItAction
+from moveit_action_handlers.msg import MoveToJointsMoveItGoal
+
+from sensor_msgs.msg import JointState
+
+# cartesian
+from moveit_action_handlers.msg import MoveToPoseMoveItAction
+from moveit_action_handlers.msg import MoveToPoseMoveItGoal
+from moveit_action_handlers.msg import PoseStamped
+from geometry_msgs.msg import Pose
+
+# forward kinematics
+# from urdf_parser_py.urdf import URDF
+# from pykdl_utils.kdl_kinematics import KDLKinematics
+
+right_gripper_joint_names = ["right_gripper_finger1_joint",
+                             "right_gripper_finger2_joint",
+                             "right_gripper_finger3_joint"]
+
+left_gripper_joint_names = ["left_gripper_finger1_joint",
+                            "left_gripper_finger2_joint",
+                            "left_gripper_finger3_joint"]
+
+
+def joint_goal(joint_values, joint_names, eev=0.5, eea=0.5, timeout=10):
+    goal = MoveToJointsMoveItGoal()
+    goal.endEffectorVelocity = eev
+    goal.endEffectorAcceleration = eea
+    goal.timeoutSeconds = timeout
+    for i in range(len(joint_names)):
+        joint_pair = PropertyValuePair()
+        joint_pair.name = joint_names[i]
+        joint_pair.value = joint_values[i]
+        goal.joint_pairs.append(joint_pair)
+    return goal
+
+
+def cartesian_goal(x, y, z, ref_link, roll=0, pitch=0, yaw=0, eev=0.5, eea=0.5, mode=0, timeout=20):
+    goal = MoveToPoseMoveItGoal()
+    goal.constraint_mode = mode
+    goal.endEffectorVelocity = eev
+    goal.endEffectorAcceleration = eea
+    goal.timeoutSeconds = timeout
+    target_pose = PoseStamped()
+    target_pose.header.frame_id = ref_link
+    target_pose.pose.position.x = x
+    target_pose.pose.position.y = y
+    target_pose.pose.position.z = z
+    target_pose.pose.orientation.roll = roll
+    target_pose.pose.orientation.pitch = pitch
+    target_pose.pose.orientation.yaw = yaw
+    goal.target_pose = target_pose
+    return goal
+
+
+class ManipulationJointControl:
+    def __init__(self):
+        self.left_arm = actionlib.simple_action_client.SimpleActionClient(
+            "/moveit_action_handlers/left_arm/joint_ctrl", MoveToJointsMoveItAction)
+        self.left_arm.wait_for_server()
```

Comment:
i recommend putting logs around these "wait for servers". too many times, my code has just hung and i have no idea why nothing is happening and it turns out i'm just waiting for a server

Reply:
Theres logs around the initialization of these objects that do this